source/simulator/introspection.cc

```diff
@@ -333,6 +333,46 @@ namespace aspect
 
 
 
+  template <int dim>
+  std::vector<unsigned int>
+  Introspection<dim>::get_compositional_field_base_element_indices() const
+  {
+    // We are assigning base elements in order, so the first compositional field
+    // gives us the first base element index. Then we find the largest index
+    // in the vector. This is necessary, because the fields could have type A,B,A.
+    std::vector<unsigned int> result;
+    const unsigned int first = this->base_elements.compositional_fields[0];
+    const unsigned int last = *std::max_element(this->base_elements.compositional_fields.begin(), this->base_elements.compositional_fields.end());
+    for (unsigned int i = first; i<=last; ++i)
+      result.emplace_back(i);
+    return result;
+  }
+
+
+
+  template <int dim>
+  std::vector<unsigned int>
+  Introspection<dim>::get_compositional_field_indices_with_base_element(const unsigned int base_element_index) const
+  {
+    std::vector<unsigned int> result;
+    Assert(base_element_index <= get_compositional_field_base_element_indices().back(),
+           ExcMessage("Invalid base_element_index specified."));
+
+    unsigned int idx = 0;
+    for (const auto base_idx : this->base_elements.compositional_fields)
+      {
+        if (base_idx == base_element_index)
+          result.emplace_back(idx);
+
+        ++idx;
+      }
```

Comment:
Also here, I think it would be better if this could be computed in the constructor once and then just returned by reference (it is called from a loop over all base elements x over all cells). I think of `introspection` as a sort of cache for internal information so it would be ok to precompute things.

Reply:
are you ok with a std::map? See the last commit.

Reply:
I am ok with it until it shows up in a profiling run ;-) ... we can keep it for now. But can you change both functions to return references to vectors? Also I think you need to indent.